filters/builtin/compress.go

```diff
@@ -104,47 +124,63 @@ func (e encodings) Swap(i, j int)      { e[i], e[j] = e[j], e[i] }
 //
 // It is possible to control the compression level, by setting it as the first
 // filter argument, in front of the MIME types. The default compression level is
-// best-speed. The possible values are integers between 0 and 9 (inclusive), where
-// 0 means no-compression, 1 means best-speed and 9 means best-compression.
+// best-speed. The possible values are integers between 0 and 11 (inclusive), where
+// 0 means no-compression, 1 means best-speed and 11 means best-compression.
 // Example:
 //
 // 	* -> compress(9, "image/tiff") -> "https://www.example.org"
 //
 // The filter also checks the incoming request, if it accepts the supported
 // encodings, explicitly stated in the Accept-Encoding header. The filter currently
-// supports gzip and deflate. It does not assume that the client accepts any
-// encoding if the Accept-Encoding header is not set. It ignores * in the
+// supports brotli, gzip and deflate. It does not assume that the client accepts any
+// encodingPriority if the Accept-Encoding header is not set. It ignores * in the
 // Accept-Encoding header.
 //
+// Supported encodings are prioritized on:
+// - quality value if provided by client
+// - server side priority (encodingPriority) otherwise
+//
 // When compressing the response, it updates the response header. It deletes the
 // the Content-Length value triggering the proxy to always return the response
-// with chunked transfer encoding, sets the Content-Encoding to the selected
-// encoding and sets the Vary: Accept-Encoding header, if missing.
+// with chunked transfer encodingPriority, sets the Content-Encoding to the selected
+// encodingPriority and sets the Vary: Accept-Encoding header, if missing.
 //
 // The compression happens in a streaming way, using only a small internal buffer.
 //
-func NewCompress() filters.Spec { return &compress{} }
+func NewCompress() filters.Spec {
+	return NewCompressWithOptions(CompressOptions{supportedEncodings})
+}
+
+func NewCompressWithOptions(options CompressOptions) filters.Spec {
+	m := map[string]int{}
+	for i, v := range options.Encodings {
+		if !stringsContain(supportedEncodings, v) {
+			log.Warningf("Skipping unsupported encoding: %s", v)
+			continue
+		}
```

Comment:
Options are derived from user-supplied string flag, by returning a error and handling it in the startup routine we would prevent Skipper from starting with unsupported values e.g. due to typo and let users notice the problem right away instead of postponing it until they ever look into the logs.

Reply:
the options are already restricted in config.go `commaListFlag("gzip", "deflate", "br")`
=> the user will get an error if he tries to use a wrong option
=> the use case would be if a developer add a new option in `commaListFlag`

I understand your concern but given that this could occur only from developer mistake, maybe we can keep it simple here in place of introducing possible error returned from new?

Reply:
Ok, fair enough, I misinterpreted `commaListFlag("gzip", "deflate", "br")` as a default value while it is actually an allowlist.

Reply:
OTOH looking forward for new options we might not be so lucky to never have to return error and adding an error would be a breaking change for the `NewCompressWithOptions`

Reply:
We have to return `filters.Spec`, there is no error that we can return. Errors should be handled upfront before calling.

Reply:
> We have to return filters.Spec

Why? We can return `filters.Spec, error` from `NewCompressWithOptions` and let caller bail out on error. The existing `NewCompress` should continue to return just a spec as it uses a valid option value.

Reply:
even though there is 99% chance that it will never happen, ignoring error on NewCompress seems a bit of a trap for developer no? :)

Reply:
Hmm maybe you are right, but I think we have no case were we have 2 return values to instantiate a spec.
I wonder if I forgot why...

Reply:
>  ignoring error on NewCompress seems a bit of a trap

We can not change `NewCompress` signature to return error due to backwards compatibility but since it does not receive any argument we can make sure it passes safe defaults to `NewCompressWithOptions` that never cause an error.

Reply:
good point, updated following above suggestions.
NewCompress is indeed tested already in the unit test